pinot-core/src/main/java/org/apache/pinot/core/data/manager/realtime/ConsumptionDelayTracker.java

```diff
@@ -0,0 +1,361 @@
+    _partitionsMarkedForVerification.remove(partitionGroupId);
+    if (_enablePerPartitionMetric) {
+      _serverMetrics.removeTableGauge(getPerPartitionMetricName(partitionGroupId),
+          ServerGauge.PER_PARTITION_CONSUMPTION_DELAY_MS);
+    }
+  }
+
+  /*
+   * Helper function to generate a per partition metric name.
+   *
+   * @param partitionGroupId the partition group id to be appended to the table name so we
+   *        can differentiate between metrics for various partitions.
+   *
+   * @return a metric name with the following structure: tableNameWithType + partitionGroupId
+   */
+  private String getPerPartitionMetricName(int partitionGroupId) {
+    return _tableNameWithType + partitionGroupId;
+  }
+
+  // Custom Constructor
+  public ConsumptionDelayTracker(ServerMetrics serverMetrics, String tableNameWithType,
+      RealtimeTableDataManager realtimeTableDataManager, int timerThreadTickIntervalMs, String metricNamePrefix,
+      boolean enableAggregateMetric, boolean enablePerPartitionMetric)
+      throws RuntimeException {
+    _logger = LoggerFactory.getLogger(tableNameWithType + "-" + getClass().getSimpleName());
+    _serverMetrics = serverMetrics;
+    _tableNameWithType = metricNamePrefix + tableNameWithType;
+    _realTimeTableDataManager = realtimeTableDataManager;
+    // Handle negative timer values
+    if (timerThreadTickIntervalMs <= 0) {
+      throw new RuntimeException("Illegal timer timeout argument, expected > 0, got=" + timerThreadTickIntervalMs);
+    }
+    _enableAging = true;
+    _enablePerPartitionMetric = enablePerPartitionMetric;
+    _enableAggregateMetric = enableAggregateMetric;
+    _timerThreadTickIntervalMs = timerThreadTickIntervalMs;
+    _timer = new Timer("ConsumptionDelayTimerThread" + tableNameWithType);
+    _timer.schedule(new TrackingTimerTask(this), INITIAL_TIMER_THREAD_DELAY_MS, _timerThreadTickIntervalMs);
+    // Install callback metric
+    if (_enableAggregateMetric) {
+      _serverMetrics.addCallbackTableGaugeIfNeeded(_tableNameWithType, ServerGauge.MAX_CONSUMPTION_DELAY_MS,
+          () -> (long) getMaxConsumptionDelay());
+    }
+  }
+
+  // Constructor that uses defaults
+  public ConsumptionDelayTracker(ServerMetrics serverMetrics, String tableNameWithType,
+      RealtimeTableDataManager tableDataManager) {
+    this(serverMetrics, tableNameWithType, tableDataManager, TIMER_THREAD_TICK_INTERVAL_MS,
+        "", true, true);
+  }
+
+  // Constructor that takes a prefix to name the metric, so we can keep multiple trackers for the same table
+  public ConsumptionDelayTracker(ServerMetrics serverMetrics, String tableNameWithType, String metricNamePrefix,
+      RealtimeTableDataManager tableDataManager) {
+    this(serverMetrics, metricNamePrefix + tableNameWithType, tableDataManager,
+        TIMER_THREAD_TICK_INTERVAL_MS, metricNamePrefix, true, true);
+  }
+
+
+  /**
+   * Use to set or rest the aging of reported values.
```

Comment:
Let us remove this call and adjust tests accordingly.

Reply:
Being tracked in another open comment, please do not duplicate